src/hmem_cache.c

```diff
@@ -0,0 +1,284 @@
+		ret = ofi_hmem_close_handle(region->key.iface, region->mapped_addr);
+		if (ret) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+					 "failed to unmap addr:%p\n", region->mapped_addr);
+		}
+
+		free(region);
+	}
+
+	FI_INFO(&core_prov, FI_LOG_CORE,
+			"%s: ipc cache purged\n", cache->name);
+}
+
+static void ipc_cache_invalidate_regions(struct hmem_cache *cache,
+										 void *from, void *to)
+{
+	struct dlist_entry region_list;
+	struct ipc_cache_region *region;
+	struct dlist_entry *tmp;
+	int ret;
+
+	dlist_init(&region_list);
+	pgtable_search_range(&cache->pgtable, (pgt_addr_t)from,
+						(pgt_addr_t)to - 1,
+						ipc_cache_region_collect_callback,
+						&region_list);
+	dlist_foreach_container_safe(&region_list, struct ipc_cache_region,
+								region, list, tmp) {
+		ret = pgtable_remove(&cache->pgtable, &region->super);
+		if (ret) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+					 "failed to remove address:%p from cache (%d)\n",
+					 (void *)region->key.address, ret);
+		}
+
+		ret = ofi_hmem_close_handle(region->key.iface, region->mapped_addr);
+		if (ret) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+					 "failed to unmap addr:%p\n", region->mapped_addr);
+		}
+		free(region);
+	}
+	FI_INFO(&core_prov, FI_LOG_CORE,
+			"%s: closed memhandles in the range [%p..%p]\n",
+			cache->name, from, to);
+}
+
+int ipc_cache_map_memhandle(struct hmem_cache *cache, struct ipc_info *key,
+							void **mapped_addr)
+{
+	pgt_region_t *pgt_region;
+	struct ipc_cache_region *region;
+	int ret;
+
+	pthread_rwlock_rdlock(&cache->lock);
+	pgt_region = pgtable_lookup(&cache->pgtable, key->address);
+	if (OFI_LIKELY(pgt_region != NULL)) {
+		region = container_of(pgt_region, struct ipc_cache_region, super);
+		if (memcmp(&key->ipc_handle, &region->key.ipc_handle,
+				   IPC_HANDLE_SIZE) == 0) {
+			/*cache hit */
+			FI_INFO(&core_prov, FI_LOG_CORE,
+					"%s: ipc cache hit addr:%p size:%lu region:"
+					PGT_REGION_FMT"\n", cache->name, (void *)key->address,
+					key->base_length, PGT_REGION_ARG(&region->super));
+
+			*mapped_addr = region->mapped_addr;
+			pthread_rwlock_unlock(&cache->lock);
+			return 0;
+		} else {
+			FI_INFO(&core_prov, FI_LOG_CORE,
+					"%s: ipc cache remove stale region:"
+					PGT_REGION_FMT  " new_addr:%p new_size:%lu\n",
+					cache->name, PGT_REGION_ARG(&region->super),
+					(void *)key->address, key->base_length);
+
+			ret = pgtable_remove(&cache->pgtable, &region->super);
+			if (ret) {
+				FI_WARN(&core_prov, FI_LOG_CORE,
+						 "%s: failed to remove address:%p from cache\n",
+						 cache->name, (void *)key->address);
+				goto err;
+			}
+
+			ret = ofi_hmem_close_handle(key->iface, region->mapped_addr);
+			if (ret) {
+				FI_WARN(&core_prov, FI_LOG_CORE,
+						 "failed to unmap addr:%p\n", region->mapped_addr);
+			}
+
+			free(region);
+		}
+	}
+
+	ret = ofi_hmem_open_handle(key->iface, (void**) &key->ipc_handle,
+							   key->base_length, key->dev_num, mapped_addr);
+	if (OFI_UNLIKELY(ret)) {
```

Comment:
If the error is the address already mapped, do you need to unmap all overlapping regions and retry? Basically what ucx does here: https://github.com/openucx/ucx/blob/master/src/uct/cuda/cuda_ipc/cuda_ipc_cache.c#L278-L310

I am hitting the same error when sender gets the same ipc handle but with a different address.